feat(citations.models): create model UnmatchedCitationFromRECAP

grossir opened the pull request:
Solves #5492

- Creates new model UnmatchedCitationFromRECAP, to be used in `recap_citations.store_recap_citations`
- add migrations for that model
- Creates an UnmatchedCitationBase model to re use code
- refactors `unmatched_citations_utils` methods to work with a RecapDocument as a citing object, from working only with Opinion objects
- adds tests
- update signal on Citation creation to update UnmatchedCitationFromRECAP status to FOUND

grossir committed:
feat(citations.models): create model UnmatchedCitationFromRECAP

Solves #5492

- Creates new model UnmatchedCitationFromRECAP, to be used in `recap_citations.store_recap_citations`
- add migrations for that model
- Creates an UnmatchedCitationBase model to re use code
- refactors `unmatched_citations_utils` methods to work with a RecapDocument as a citing object, from working only with Opinion objects
- adds tests
- update signal on Citation creation to update UnmatchedCitationFromRECAP status to FOUND

grossir committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

grossir commented:
Just realized we don't update `OpinionCluster.citation_count` on the `store_recap_citations` function... should we?
This may be another way in which that field is not accurate
https://github.com/freelawproject/courtlistener/issues/5601

grossir committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

grossir committed:
fix(UnmatchedCitationFromRecap): update migrations

Bug fixes
- unmatched_citations_utils.py: bulk_create now uses the dynamic unmatched_citation_model (merge resolution had reverted to hardcoded UnmatchedCitation, sending RECAP rows
to the wrong table); kept main's ignore_conflicts=True
- tests.py: renamed OpinionClusterFactoryWithChildrenAndParents → OpinionClusterWithChildrenAndParentsFactory (renamed on main); removed unused citation2 = assignment

Migrations (fixes the CI failure: conflicting 0002 leaf nodes)
- Deleted stale 0002_add_unmatched_citation_from_recap.py and misnamed 002_...sql
- Regenerated as 0003_add_unmatched_citation_from_recap.py + .sql via docker makemigrations/sqlmigrate — now depends on 0002_alter_unmatchedcitation_volume, volume is
TextField (old file had smallint)

Polish
- models.py: proper docstrings on both create_from_eyecite overrides (copy-pasted/wrong params before); return type annotations (Self, explicit classes); removed stray #
comments in Meta
- unmatched_citations_utils.py: Sequence[BaseUnmatchedCitation] type hint; fixed docstring param names (citing_object, resolved_citations); removed redundant .all()
- signals.py: fixed signal docstring ("status status to MATCHED" → FOUND, mentions both models)

grossir committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

quevon24 committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

quevon24 committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

quevon24 commented on cl/citations/models.py:
```diff
@@ -1,56 +1,54 @@
+class BaseUnmatchedCitation(BaseCitation):
+    """
+    Track citations that couldn't be resolved to a RecapDocument or
+    OpinionCluster
```
Opinion instead of OpinionCluster? citing objects point to Opinion or RECAPDocument

grossir replied:
added this to the new commit

quevon24 commented on cl/citations/models.py:
```diff
@@ -124,3 +115,89 @@ def create_from_eyecite(
+        return unmatched_citation
+
+
+class UnmatchedCitationFromRECAPDocument(BaseUnmatchedCitation):
```
Register the new model in admin? As is being done with UnmatchedCitation

grossir replied:
added this to the new commit

quevon24 commented on cl/citations/tests.py:
```diff
@@ -629,18 +632,27 @@ def test_unsafe_case_names(self) -> None:
         )
         # Courts
-        court_ca2 = CourtFactory(id="ca2")
+        court_ca2 = cls.court_ca2 = CourtFactory(id="ca2")
```
why this double assignment?

grossir replied:
it's just because court_ca2 is used again below in that same func; but we also need to reference it in other funcs. So both names are used

quevon24 commented on cl/citations/models.py:
```diff
@@ -124,3 +115,89 @@ def create_from_eyecite(
+            )
+        ]
+        unique_together = (
+            ("citing_recapdocument", "volume", "reporter", "page"),
```
Why dont use UniqueConstraint instead? this is a new model

grossir replied:
added this to the new commit

quevon24 commented on cl/citations/unmatched_citations_utils.py:
```diff
@@ -49,62 +54,71 @@ def unmatched_citation_is_valid(
             ]:
                 continue
-            found.status = UnmatchedCitation.FAILED
+            found.status = BaseUnmatchedCitation.FAILED
```
Cant we do this in bulk?

grossir replied:
yes, I added this to the new commit

quevon24 reviewed (commented):
Just a few comments, nothing blocks the merging of this PR

grossir committed:
refactor(citations): address review feedback on UnmatchedCitationFromRECAPDocument

- Reword BaseUnmatchedCitation docstring to describe citing Opinion/RECAPDocument resolved to a cluster.
- Register UnmatchedCitationFromRECAPDocument in admin via a shared UnmatchedCitationAdminMixin.
- Replace the new model's unique_together with a named UniqueConstraint.
- Use bulk_update instead of per-row save() in update_unmatched_citations_status.

grossir committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

quevon24 committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

quevon24 reviewed (approved):
Thanks for addressing my comments, It is ready.

ERosendo committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

chaco-fl committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap

chaco-fl committed:
Merge branch 'main' into 5492-unmatched-citations-from-recap
